commit 4f2ba4ecdad0a39fa336ee927bc4eddcd1b17eba
Author: Eric Anholt <eric@anholt.net>
Date:   Tue Aug 15 16:47:19 2017 -0700

    drm/vc4: Avoid using vrefresh==0 mode in DSI htotal math.
    
    The incoming mode might have a missing vrefresh field if it came from
    drmModeSetCrtc(), which the kernel is supposed to calculate using
    drm_mode_vrefresh().  We could either use that or the adjusted_mode's
    original vrefresh value.
    
    However, we can maintain a more exact vrefresh value (not just the
    integer approximation), by scaling by the ratio of our clocks.
    
    v2: Use math suggested by Andrzej Hajda instead.
    v3: Simplify math now that adjusted_mode->clock isn't padded.
    v4: Drop some parens.
    
    Signed-off-by: Eric Anholt <eric@anholt.net>
    Link: https://patchwork.freedesktop.org/patch/msgid/20170815234722.20700-2-eric@anholt.net
    Reviewed-by: Andrzej Hajda <a.hajda@samsung.com>
    (cherry picked from commit af2eca53206c59ce9308a4f5f46c4a104a179b6b)

diff --git a/drivers/gpu/drm/vc4/vc4_dsi.c b/drivers/gpu/drm/vc4/vc4_dsi.c
index 1b71ebb66b50..28731b23186a 100644
--- a/drivers/gpu/drm/vc4/vc4_dsi.c
+++ b/drivers/gpu/drm/vc4/vc4_dsi.c
@@ -955,7 +955,8 @@ static bool vc4_dsi_encoder_mode_fixup(struct drm_encoder *encoder,
 	adjusted_mode->clock = pixel_clock_hz / 1000 + 1;
 
 	/* Given the new pixel clock, adjust HFP to keep vrefresh the same. */
-	adjusted_mode->htotal = pixel_clock_hz / (mode->vrefresh * mode->vtotal);
+	adjusted_mode->htotal = adjusted_mode->clock * mode->htotal /
+				mode->clock;
 	adjusted_mode->hsync_end += adjusted_mode->htotal - mode->htotal;
 	adjusted_mode->hsync_start += adjusted_mode->htotal - mode->htotal;
 
